### ISSUE TITLE
[R-package] predict() always predicts 0.0 if using a custom objective + target with missing values

### ISSUE THREAD

**Issue by jameslamb** [microsoft/LightGBM#4316]
## Reproducible example

Running the following R code, you'll see that `preds` is all 0s.

```r
library(data.table)
library(lightgbm)
library(nycflights13)

.l2_loss <- function(preds, dtrain) {
    labels <- getinfo(dtrain, "label")
    print(head(preds))
    grad <- preds - labels
    hess <- rep(1.0, length(preds))
    print(head(grad))
    return(list(grad = grad, hess = hess))
}

weatherDT <- data.table::as.data.table(nycflights13::weather)

# if you comment out this line, there is one NA in y_train and no trees are added
# weatherDT <- weatherDT[!is.na(humid)]

# convert from data frame to matrix
feature_names <- c(
    "month"
    , "day"
    , "hour"
    , "temp"
    , "dewp"
    , "wind_dir"
    , "wind_speed"
    , "precip"
    , "pressure"
)
target_name <- "humid"
X_train <- data.matrix(weatherDT[ month <= 11, .SD, .SDcols = feature_names])
y_train <- weatherDT[month <= 11,][[target_name]]
X_test <- data.matrix(weatherDT[ month == 12, .SD, .SDcols = feature_names])
y_test <- weatherDT[month == 12,][[target_name]]

dtrain <- lightgbm::lgb.Dataset(
    data = X_train
    , label = y_train
    , colnames = feature_names
    , categorical_feature = c("month", "day", "hour")
)

mod <- lightgbm::lgb.train(
    data = dtrain
    , nrounds = 5L
    , params = list(
        min_gain_to_split = 0.0
        , min_data_in_leaf = 1L
        , boost_from_average = FALSE
    )
    , objective = .l2_loss
)
preds <- predict(mod, X_test)
str(preds)
```

> [LightGBM] [Warning] Auto-choosing col-wise multi-threading, the overhead of testing was 0.002126 seconds.
You can set `force_col_wise=true` to remove the overhead.
[LightGBM] [Info] Total Bins 690
[LightGBM] [Info] Number of data points in the train set: 23971, number of used features: 9
[LightGBM] [Warning] Using self-defined objective function
[1] 0 0 0 0 0 0
[1] -59.37 -61.63 -64.43 -62.21 -64.43 -67.21
[LightGBM] [Warning] No further splits with positive gain, best gain: -inf
[LightGBM] [Warning] Stopped training because there are no more leaves that meet the split requirements
[1] 0 0 0 0 0 0
[1] -59.37 -61.63 -64.43 -62.21 -64.43 -67.21
[LightGBM] [Warning] No further splits with positive gain, best gain: -inf
[LightGBM] [Warning] Stopped training because there are no more leaves that meet the split requirements
[1] 0 0 0 0 0 0
[1] -59.37 -61.63 -64.43 -62.21 -64.43 -67.21
[LightGBM] [Warning] No further splits with positive gain, best gain: -inf
[LightGBM] [Warning] Stopped training because there are no more leaves that meet the split requirements
[1] 0 0 0 0 0 0
[1] -59.37 -61.63 -64.43 -62.21 -64.43 -67.21
[LightGBM] [Warning] No further splits with positive gain, best gain: -inf
[LightGBM] [Warning] Stopped training because there are no more leaves that meet the split requirements
[1] 0 0 0 0 0 0
[1] -59.37 -61.63 -64.43 -62.21 -64.43 -67.21
[LightGBM] [Warning] No further splits with positive gain, best gain: -inf
[LightGBM] [Warning] Stopped training because there are no more leaves that meet the split requirements

>  num [1:2144] 0 0 0 0 0 0 0 0 0 0 ...

If you the same code is run but with `objective` changed to `"regression_l2"` or any of the other LightGBM builtins, training will succeed and produce non-constant predictions.

> [LightGBM] [Warning] Auto-choosing row-wise multi-threading, the overhead of testing was 0.000631 seconds.
You can set `force_row_wise=true` to remove the overhead.
And if memory is not enough, you can set `force_col_wise=true`.
[LightGBM] [Info] Total Bins 690
[LightGBM] [Info] Number of data points in the train set: 23971, number of used features: 9

If the test code is run with that custom objective, but without any missing values in the target, training will succeed and produce non-constant predictions.

> [LightGBM] [Warning] Using self-defined objective function
[LightGBM] [Warning] Auto-choosing row-wise multi-threading, the overhead of testing was 0.000644 seconds.
You can set `force_row_wise=true` to remove the overhead.
And if memory is not enough, you can set `force_col_wise=true`.
[LightGBM] [Info] Total Bins 688
[LightGBM] [Info] Number of data points in the train set: 23970, number of used features: 9
[LightGBM] [Warning] Using self-defined objective function
[1] 0 0 0 0 0 0
[1] -59.37 -61.63 -64.43 -62.21 -64.43 -67.21
[1] 5.571214 5.571214 5.571214 5.571214 5.571214 5.571214
[1] -53.79879 -56.05879 -58.85879 -56.63879 -58.85879 -61.63879
[1] 10.46168 10.46168 10.46168 10.46168 10.46168 10.46168
[1] -48.90832 -51.16832 -53.96832 -51.74832 -53.96832 -56.74832
[1] 15.1626 15.1626 15.1626 15.1626 15.1626 15.1626
[1] -44.2074 -46.4674 -49.2674 -47.0474 -49.2674 -52.0474
[1] 19.47322 19.47322 19.47322 19.47322 19.47322 19.47322
[1] -39.89678 -42.15678 -44.95678 -42.73678 -44.95678 -47.7367

>  num [1:2144] 26 25.1 24.6 26 33.5 ...

## Environment info

LightGBM version or commit hash: latest `master` (https://github.com/microsoft/LightGBM/commit/da3465cbf1d1cf43e0d14908f110814e960905bc)

Command(s) you used to install LightGBM

```shell
sh build-cran-package.sh
R CMD INSTALL lightgbm_3.2.1.99.tar.gz
```

<details><summary>output of 'sessionInfo()'</summary>

```text
R version 4.0.5 (2021-03-31)
Platform: x86_64-apple-darwin17.0 (64-bit)
Running under: macOS Mojave 10.14.6

Matrix products: default
BLAS:   /System/Library/Frameworks/Accelerate.framework/Versions/A/Frameworks/vecLib.framework/Versions/A/libBLAS.dylib
LAPACK: /Library/Frameworks/R.framework/Versions/4.0/Resources/lib/libRlapack.dylib

locale:
[1] en_US.UTF-8/en_US.UTF-8/en_US.UTF-8/C/en_US.UTF-8/en_US.UTF-8

attached base packages:
[1] stats     graphics  grDevices utils     datasets  methods   base     

other attached packages:
[1] nycflights13_1.0.2 lightgbm_3.2.1.99  R6_2.4.1           data.table_1.14.0 

loaded via a namespace (and not attached):
 [1] lattice_0.20-41 fansi_0.4.1     crayon_1.4.1    utf8_1.1.4     
 [5] grid_4.0.5      lifecycle_0.2.0 jsonlite_1.7.2  magrittr_2.0.1 
 [9] pillar_1.5.1    rlang_0.4.10    Matrix_1.3-3    vctrs_0.3.6    
[13] ellipsis_0.3.1  tools_4.0.5     compiler_4.0.5  pkgconfig_2.0.3
[17] tibble_3.1.0
```

</details>

## Additional Comments

I have not investigated this further yet, just wanted to document it since I noticed this while prepping for a talk.

Steps that could help find the root cause:

* simplifying the reproducible example
* checking the behavior of the Python package when there are missing values of the target

**Comment by jameslamb** [comment]
ooo interesting. I just had an idea after writing this up. Maybe the issue is just that my custom loss function doesn't explicitly handle missing values.

If I change it to the following

```r
.l2_loss <- function(preds, dtrain) {
    labels <- getinfo(dtrain, "label")
    print(head(preds))
    grad <- preds - labels

    # explicitly handle missing values
    grad[is.na(grad)] <- 0.0

    hess <- rep(1.0, length(preds))
    print(head(grad))
    return(list(grad = grad, hess = hess))
}
```

then the predictions from training with my custom loss function and the builtin `regression_l2` are very close to each other.

Will investigate more in the next few days.
